dtests/src/test/java/io/snappydata/hydra/cluster/SnappyTest.java

```diff
@@ -1466,13 +1539,24 @@ protected void executeSparkJob(Vector jobClassNames, String logFileName) {
             for (int i = 0; i < jobClassNames.size(); i++) {
                 String userJob = (String) jobClassNames.elementAt(i);
                 String masterHost = (String) SnappyBB.getBB().getSharedMap().get("masterHost");
-                String locatorHost = (String) SnappyBB.getBB().getSharedMap().get("locatorHost");
-                String command = snappyJobScript + " --class " + userJob + " --master spark://" + masterHost + ":" + MASTER_PORT + " --conf snappydata.store.locators=" + locatorHost + ":" + 10334 + " " + snappyTest.getUserAppJarLocation(userAppJar);
+                String locatorsList = getLocatorsList("locators");
+                String command = snappyJobScript + " --class " + userJob + " --master spark://" + masterHost + ":" + MASTER_PORT + " --conf snappydata.store.locators=" + locatorsList + " " + snappyTest.getUserAppJarLocation(userAppJar);
                 log = new File(".");
                 String dest = log.getCanonicalPath() + File.separator + logFileName;
                 logFile = new File(dest);
                 pb = new ProcessBuilder("/bin/bash", "-c", command);
                 snappyTest.executeProcess(pb, logFile);
+                if (SnappyPrms.waitForSparkJobCompletion()) {
+                    String searchString = SnappyPrms.getSearchStringForSparkJob();
+                    if (searchString == null)
+                        throw new TestException("Search String notifying completion of spark job is missing for job : " + userJob);
+                    String expression = "cat " + logFile + " | grep -e Exception -e '" + searchString + "' | grep -v java.net.BindException" + " | wc -l)\"";
+                    String searchCommand = "while [ \"$(" + expression + " -le  0 ] ; do sleep 1 ; done";
+                    pb = new ProcessBuilder("/bin/bash", "-c", searchCommand);
+                    Log.getLogWriter().info("spark job " + userJob + " starts at: " + System.currentTimeMillis());
+                    executeProcess(pb, logFile);
+                    Log.getLogWriter().info("spark job  " + userJob + " finishes at:  " + System.currentTimeMillis());
+                }
```

Comment:
everything can be avoided if we can register a  SparkListener . see above comment